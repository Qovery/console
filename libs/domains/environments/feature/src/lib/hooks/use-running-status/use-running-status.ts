import { useQuery } from '@tanstack/react-query'
import { queries } from '@qovery/state/util-queries'

export interface UseRunningStatusProps {
  environmentId?: string
}

export function useRunningStatus({ environmentId }: UseRunningStatusProps) {
  return useQuery({
    // eslint-disable-next-line @typescript-eslint/no-extra-non-null-assertion
    ...queries.environments.runningStatus(environmentId!!),
    enabled: Boolean(environmentId),
    refetchOnMount: false,
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
    staleTime: Infinity,
  })
}

export default useRunningStatus
