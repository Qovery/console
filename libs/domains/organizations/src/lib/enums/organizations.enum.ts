export enum OrganizationPlanType {
  COMMUNITY = 'COMMUNITY',
  FREE = 'FREE',
  PROFESSIONAL = 'PROFESSIONAL',
  STARTUP = 'STARTUP',
  ENTERPRISE = 'ENTERPRISE',
  BUSINESS = 'BUSINESS',
}
