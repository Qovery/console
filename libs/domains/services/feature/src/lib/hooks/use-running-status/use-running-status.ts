import { useQuery } from '@tanstack/react-query'
import { queries } from '@qovery/state/util-queries'

export interface UseRunningStatusProps {
  environmentId?: string
  serviceId?: string
}

export function useRunningStatus({ environmentId, serviceId }: UseRunningStatusProps) {
  return useQuery({
    // eslint-disable-next-line @typescript-eslint/no-extra-non-null-assertion
    ...queries.services.runningStatus(environmentId!!, serviceId!!),
    enabled: Boolean(environmentId) && Boolean(serviceId),
    refetchOnMount: false,
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
    staleTime: Infinity,
  })
}

export default useRunningStatus
