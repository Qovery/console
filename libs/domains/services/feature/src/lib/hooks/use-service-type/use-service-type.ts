import { useQuery } from '@tanstack/react-query'
import { queries } from '@qovery/state/util-queries'

export interface UseServiceTypeProps {
  environmentId?: string
  serviceId?: string
}

export function useServiceType({ environmentId, serviceId }: UseServiceTypeProps) {
  return useQuery({
    ...queries.services.list(environmentId!),
    select(data) {
      return data.find(({ id }) => id === serviceId)?.serviceType
    },
    enabled: Boolean(environmentId),
  })
}

export default useServiceType
