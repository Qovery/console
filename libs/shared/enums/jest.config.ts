/* eslint-disable */
export default {
  displayName: 'shared-enums',
  preset: '../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': ['babel-jest', { presets: ['@nrwl/react/babel'] }],
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../coverage/libs/shared/enums',
}
