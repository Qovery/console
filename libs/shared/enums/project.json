{
  "sourceRoot": "libs/shared/enums/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/shared/enums/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/shared/enums"],
      "options": {
        "jestConfig": "libs/shared/enums/jest.config.ts",
        "passWithNoTests": true
      }
    }
  }
}
