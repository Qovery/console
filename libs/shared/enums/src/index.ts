export * from './lib/colors.enum'
export * from './lib/service-type.enum'
export * from './lib/running-status.enum'
export * from './lib/logs-type.enum'
export * from './lib/memory-size.enum'
export * from './lib/job.type'
export * from './lib/utils/get-service-type'
export * from './lib/icon.enum'
