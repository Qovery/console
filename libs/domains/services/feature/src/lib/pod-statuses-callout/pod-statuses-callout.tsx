import { type IconName } from '@fortawesome/fontawesome-common-types'
import { type ReactNode, useState } from 'react'
import { useNavigate, useParams } from 'react-router-dom'
import { APPLICATION_SETTINGS_DOMAIN_URL, APPLICATION_SETTINGS_URL, APPLICATION_URL } from '@qovery/shared/routes'
import { Button, Callout, type CalloutRootProps, Icon, IconAwesomeEnum } from '@qovery/shared/ui'
import { upperCaseFirstLetter } from '@qovery/shared/util-js'
import { useCleanFailedJobs } from '../hooks/use-clean-failed-jobs/use-clean-failed-jobs'
import { useRunningStatus } from '../hooks/use-running-status/use-running-status'
import { useServiceType } from '../hooks/use-service-type/use-service-type'

export interface PodStatusesCalloutProps {
  environmentId: string
  serviceId: string
}

export function PodStatusesCallout({ environmentId, serviceId }: PodStatusesCalloutProps) {
  const navigate = useNavigate()
  const { organizationId = '', projectId = '' } = useParams()
  const { data: runningStatuses, isLoading: isRunningStatusesLoading } = useRunningStatus({ environmentId, serviceId })
  const { data: serviceType, isLoading: isServiceTypeLoading } = useServiceType({ environmentId, serviceId })
  const { mutate: cleanFailedJobs } = useCleanFailedJobs()
  const [activeIndex, setActiveIndex] = useState(0)

  if (isRunningStatusesLoading || !runningStatuses || isServiceTypeLoading) {
    return null
  }

  const callouts: {
    id: number
    icon: IconName
    color: CalloutRootProps['color']
    title: string
    description?: ReactNode
    children?: ReactNode
  }[] = []

  if (runningStatuses.state === 'ERROR') {
    callouts.push({
      id: 1,
      icon: 'circle-exclamation',
      color: 'red',
      title: serviceType === 'JOB' ? 'Job execution failure' : 'Application pods are in error',
      description:
        serviceType === 'JOB'
          ? 'One of the job executions have failed. Have a look at the table below for investigation.'
          : 'Some pods are experiencing issues. Have a look at the table below for investigation.',
      children:
        serviceType === 'JOB' && environmentId && serviceId ? (
          <div className="flex flex-row items-center">
            <Button
              type="button"
              color="neutral"
              variant="outline"
              size="md"
              onClick={() => cleanFailedJobs({ environmentId, payload: { job_ids: [serviceId] } })}
            >
              Clear status
            </Button>
          </div>
        ) : undefined,
    })
  }

  if (runningStatuses.state === 'WARNING' || runningStatuses.state === 'ERROR') {
    const certificatesInError =
      'certificates' in runningStatuses && runningStatuses.certificates.filter(({ state }) => state === 'ERROR')
    if (certificatesInError && certificatesInError.length > 0) {
      callouts.push({
        id: 3,
        icon: 'circle-exclamation',
        color: 'red',
        title: 'Certificate Issues',
        description:
          'One or more certificates couldnâ€™t be generated for your domains. Have a look at the domain section to know more.',
        children: (
          <div className="flex flex-row items-center">
            <Button
              type="button"
              color="neutral"
              variant="outline"
              className="text-md gap-2"
              size="md"
              onClick={() =>
                navigate(
                  APPLICATION_URL(organizationId, projectId, environmentId, serviceId) +
                    APPLICATION_SETTINGS_URL +
                    APPLICATION_SETTINGS_DOMAIN_URL
                )
              }
            >
              Domain settings
              <Icon iconName="gear-complex" iconStyle="regular" />
            </Button>
          </div>
        ),
      })
    } else {
      callouts.push({
        id: 2,
        icon: 'circle-exclamation',
        color: 'yellow',
        title: 'Application pods have experienced issues in the past',
        description: 'Some pods experienced issues in the past. Have a look at the table below for investigation.',
        children:
          serviceType === 'JOB' && environmentId && serviceId ? (
            <div className="flex flex-row items-center">
              <Button
                type="button"
                color="neutral"
                variant="outline"
                size="md"
                onClick={() => cleanFailedJobs({ environmentId, payload: { job_ids: [serviceId] } })}
              >
                Clear status
              </Button>
            </div>
          ) : undefined,
      })
    }
  }

  if (callouts.length === 0) {
    callouts.push({
      id: 4,
      icon: 'circle-check',
      color: 'green',
      title: `Service status: ${upperCaseFirstLetter(runningStatuses.state)}`,
    })
  }

  return (
    <>
      {callouts
        .filter((_, index) => index === activeIndex)
        .map(({ id, icon, color, children, title, description: description }) => (
          <Callout.Root color={color} key={id}>
            <Callout.Icon>
              <Icon iconName={icon} iconStyle="regular" />
            </Callout.Icon>
            <Callout.Text>
              <Callout.TextHeading>
                {callouts.length > 1 ? `${activeIndex + 1}/${callouts.length} ` : undefined}
                {title}
              </Callout.TextHeading>
              <Callout.TextDescription>{description}</Callout.TextDescription>
            </Callout.Text>
            {children}
            <div className="flex flex-row items-center gap-1.5">
              {callouts.length > 1 && (
                <>
                  <Button
                    type="button"
                    color="neutral"
                    variant="outline"
                    size="md"
                    disabled={activeIndex === 0}
                    onClick={() => setActiveIndex((index) => index - 1)}
                  >
                    <Icon className="px-1" iconName="chevron-left" />
                  </Button>
                  <Button
                    type="button"
                    color="neutral"
                    variant="outline"
                    size="md"
                    disabled={activeIndex === callouts.length - 1}
                    onClick={() => setActiveIndex((index) => index + 1)}
                  >
                    <Icon className="px-1" iconName="chevron-right" />
                  </Button>
                </>
              )}
            </div>
          </Callout.Root>
        ))}
    </>
  )
}

export default PodStatusesCallout
