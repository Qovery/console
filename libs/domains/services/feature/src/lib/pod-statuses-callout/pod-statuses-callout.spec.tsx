import { ServiceStateDto } from 'qovery-ws-typescript-axios'
import { ServiceTypeEnum } from '@qovery/shared/enums'
import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import * as useRunningStatusImport from '../hooks/use-running-status/use-running-status'
import * as useServiceTypeImport from '../hooks/use-service-type/use-service-type'
import { PodStatusesCallout } from './pod-statuses-callout'

describe('PodStatusesCallout', () => {
  it('should render successfully', () => {
    jest.spyOn(useServiceTypeImport, 'useServiceType').mockReturnValue({
      data: ServiceTypeEnum.APPLICATION,
      isLoading: false,
      error: {},
    })
    jest.spyOn(useRunningStatusImport, 'useRunningStatus').mockReturnValue({
      data: {
        state: ServiceStateDto.ERROR,
        id: 'id',
        pods: [
          {
            containers: [],
            name: 'foo',
            restart_count: 0,
            service_version: '',
            state: ServiceStateDto.ERROR,
            state_message: 'bar',
            state_reason: 'baz',
          },
        ],
      },
      isLoading: false,
    })
    const { baseElement } = renderWithProviders(<PodStatusesCallout environmentId="1" serviceId="1" />)
    expect(baseElement).toMatchSnapshot()
  })
  it('should handle certificate errors properly', async () => {
    jest.spyOn(useServiceTypeImport, 'useServiceType').mockReturnValue({
      data: ServiceTypeEnum.APPLICATION,
      isLoading: false,
      error: {},
    })
    jest.spyOn(useRunningStatusImport, 'useRunningStatus').mockReturnValue({
      data: {
        state: ServiceStateDto.WARNING,
        id: 'id',
        pods: [
          {
            containers: [],
            name: 'foo2',
            restart_count: 0,
            service_version: '',
            state: ServiceStateDto.WARNING,
            state_message: 'bar',
            state_reason: 'baz',
          },
        ],
        certificates: [
          {
            dns_names: ['foo.com', 'bar.com'],
            failed_issuance_attempt_count: 0,
            state: ServiceStateDto.ERROR,
          },
        ],
      },
      isLoading: false,
    })
    renderWithProviders(<PodStatusesCallout environmentId="1" serviceId="1" />)
    const buttons = screen.getAllByRole('button')
    expect(buttons).toHaveLength(1)
  })
})
