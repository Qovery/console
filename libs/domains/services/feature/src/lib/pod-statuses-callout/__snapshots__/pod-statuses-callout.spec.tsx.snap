// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PodStatusesCallout should render successfully 1`] = `
<body>
  <div>
    <div
      class="flex flex-row gap-x-3 p-3 border rounded text-sm border-red-500 bg-red-50 text-red-500"
      data-accent-color="red"
    >
      <div
        class="-order-2 text-lg leading-6"
      >
        <i
          aria-hidden="true"
          class="fa-regular fa-circle-exclamation "
        />
      </div>
      <div
        class="-order-1 mr-auto gap-x-3 text-neutral-400"
      >
        <span
          class="block font-medium leading-6"
        >
          1/2 
          Application pods are in error
        </span>
        <span
          class="text-neutral-400"
        >
          Some pods are experiencing issues. Have a look at the table below for investigation.
        </span>
      </div>
      <div
        class="flex flex-row items-center gap-1.5"
      >
        <button
          class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none bg-white dark:bg-neutral-600 outline-brand-500 text-sm h-9 px-3 rounded border border-neutral-250 dark:border-neutral-400 text-neutral-400 dark:text-white hover:[&:not(:active)]:border-neutral-300 dark:hover:[&:not(:active):not(:disabled)]:border-neutral-350 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-100 disabled:border-none"
          disabled=""
          type="button"
        >
          <i
            aria-hidden="true"
            class="fa-solid fa-chevron-left px-1"
          />
        </button>
        <button
          class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none bg-white dark:bg-neutral-600 outline-brand-500 text-sm h-9 px-3 rounded border border-neutral-250 dark:border-neutral-400 text-neutral-400 dark:text-white hover:[&:not(:active)]:border-neutral-300 dark:hover:[&:not(:active):not(:disabled)]:border-neutral-350 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-100 disabled:border-none"
          type="button"
        >
          <i
            aria-hidden="true"
            class="fa-solid fa-chevron-right px-1"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
